jsapp/js/libs/reactCreateBemElement.d.ts

```diff
@@ -0,0 +1,13 @@
+interface BemComponent extends React.ComponentClass {
+  __: (className: string, elementType?: string) => BemComponent
+}
+
+export interface BemDefinition {
+  [elementName: string]: BemComponent
+  create: (className: string, elementType?: string) => BemComponent
```

Comment:
semi-related but are there eslint configs that we are supposed to include? mine is complaining about the syntax:
![image](https://user-images.githubusercontent.com/16762675/130129146-35fa272e-6e37-4073-acce-52a9a81457b6.png)

Reply:
The problem with types here is the `reactCreateBemElement` file is a mess :-P

Reply:
> create doesn't seem to be explicitly mentioned in the use for this interface, is it a keyword for a sort of interface constructor?
It's not a keyword or anything. When we import `BEM` (uppercase) into `bem.ts`, we create an instance with `const bem = BEM.init()`. And now this `cons bem` is keeping all the bem modules we define (e.g.g `bem.LoadingSpinner`) but also has a `create` method, that we define at the end of `bem.ts` file: `bem.create = BEM;` :-D

Reply:
> So `bem.<someRootElement> = bem('<someRootElement>')` is actually invoking this `create` method?

`bem` (lowercase) is a function that can create new bem modules and is also keeping all of them as properties. It uses `reactCreateBemElement` underneath.

`BEM` (uppercase) is also a function that can create new bem modules and is a direct alias to `reactCreateBemElement`. Currently we use either `BEM()` or `bem.create` to create new bem modules.

Reply:
> and now `bem.<someRootElement>` is of type `BemComponent`?

Yes :)

Reply:
I'm not sure what's going on with `eslint` as I did not have these issues. But for sure we will need to take a look at linting. There is `ts-lint` project that I used to use, but it's dead and depracated now :( There is a `typescript-eslint` plugin for `eslint` but I tried to configure it and had issues to run it (it wants to be the only linter plugin, and didn't work with our `es6` files, but surely there must be some complex way to have it)

Reply:
> I tried to configure it and had issues to run it (it wants to be the only linter plugin, and didn't work with our es6 files, but surely there must be some complex way to have it)

Ah, right. I think we might differ here as the tool I'm using is a plugin of a vim plugin :D

Reply:
>`bem` (lowercase) is a function that can create new bem modules and is also keeping all of them as properties. It uses `reactCreateBemElement` underneath.
>
>`BEM` (uppercase) is also a function that can create new bem modules and is a direct alias to `reactCreateBemElement`. Currently we use either `BEM()` or `bem.create` to create new bem modules.

Follow up question, why do we use `bem.create` (and create the instance) instead of directly using `BEM()` then?

Reply:
One more I forgot to ask, what is the `.d` in the file extension for?

Reply:
> Follow up question, why do we use `bem.create` (and create the instance) instead of directly using `BEM()` then?
We use `bem.create` so the bem module we're creating is being kept at `bem.` and thus being accessible. We use `BEM()` only in the `bem.ts` file and I guess we shouldn't as it's no much difference :D

Reply:
> One more I forgot to ask, what is the `.d` in the file extension for?

`d` stands for `definitions`